File: Py/day-01.py
# --- Global Configuration ---

# The current position of the lock. Starts at 50.
lock_current_state = 50
# The size of the circular lock (100 possible positions).
lock_wrap = 100


# --- Movement Functions (Calculating Final State) ---


def left_turn(n: int):
    """
    Calculates the new lock state after turning Left `n` clicks.
    """
    # The result is (Current Position - Distance) modulo Lock Size, i.e. 100.
    # Python's modulo operator (%) handles negative numbers correctly for circular arrays:
    # e.g., (50 - 60) % 100 = -10 % 100 = 90
    return (lock_current_state - n) % lock_wrap


def right_turn(n: int):
    """
    Calculates the new lock state after turning Right `n` clicks.
    """
    # The result is (Current Position + Distance) modulo Lock Size, i.e. 100.
    # e.g., (50 + 60) % 100 = 110 % 100 = 10
    return (lock_current_state + n) % lock_wrap


# --- Wrap Counting Function ---


def wraps(value, delta):
    """
    Calculates how many times the lock wraps around the `0` boundary
    given a starting `value` and a signed distance `delta`.
    This uses O(1) arithmetic instead of simulating every step.
    """
    # Calculate the theoretical position without any modulo wrap-around.
    tmp = value + delta

    if delta > 0:  # Right (R) turn
        # The number of wraps is determined by the difference in "full cycle counts".
        # Integer division (//) gives the number of times 100 fits into the number.
        # This count represents the number of full wraps passed.
        clicks = (tmp // lock_wrap) - (value // lock_wrap)
    else:  # Left (L) turn (delta is negative)
        # For negative deltas, using (value - 1) helps correct for boundary cases
        # near zero with Python's 'floor' division on negative numbers.
        clicks = ((value - 1) // lock_wrap) - ((tmp - 1) // lock_wrap)

    # Return the absolute count of wraps (always a positive number).
    return abs(clicks)


#

# --- Main Program Logic ---


def main():
    # Declare the global variable to be modified inside the function.
    global lock_current_state
    # Map direction characters from the input file to the corresponding function.
    function_array = {"R": right_turn, "L": left_turn}

    # Counter for Block 1's goal: how many times the state lands exactly on 0.
    zero_counter = 0

    # Counter for Block 2's goal: total number of times the lock wraps (crosses 0).
    wrap_counter = 0

    print(f"--- Starting state: {lock_current_state} (Wrap limit: {lock_wrap}) ---")

    try:
        # Open and process the input file.
        with open("input-2025.txt") as t:
            for line in t:
                line = line.strip()
                if not line:
                    continue

                direction = line[0].upper()
                try:
                    # Extract the numerical distance (n) from the rest of the string.
                    n = int(line[1:])
                except ValueError:
                    # skip lines where the distance is not an integer.
                    print(f"Skipping malformed line: {line} (non-integer rotation)")
                    continue

                # --- Logic for Block 2 ---

                # Determine the signed distance (delta) for the wraps function.
                delta = n if direction == "R" else -n
                # Calculate the number of wraps using the O(1) arithmetic function.
                lock_wrap_count = wraps(lock_current_state, delta)

                # --- Logic for Block 1 ---

                # Calculate the final position using the appropriate modulo function (right_turn or left_turn).
                new_state = function_array[direction](n)

                # Update the total wrap counter.
                wrap_counter += lock_wrap_count
                # Update the global state for the next instruction.
                lock_current_state = new_state

                # --- Final Checks ---

                # Check if the final position landed exactly on 0 (Part 1 Goal).
                if lock_current_state == 0:
                    zero_counter += 1

    except FileNotFoundError:
        # Inform the user if the required input file is missing.
        print(
            "\nError: 'input.txt' not found. Please create it with lines like 'L67' or 'R21'."
        )
        return  # Exit main function

    # --- Output Results ---
    print(f"Part 1 Answer: {zero_counter}")
    print(f"Part 2 Answer: {wrap_counter}")


# Standard Python entry point: ensures main() runs when the script is executed directly.
if __name__ == "__main__":
    main()

File: JS/day-01.js
"use strict"; 
/**
 * Enabling strict mode:
 * - Prevents accidental globals
 * - Throws errors for unsafe behavior
 * - Allows JavaScript engines to optimize better
 * 
 * Think of this as "no silent mistakes allowed".
 */
const { readFileSync } = require("fs");


/* -------------------------------------------------------------------------- */
/*                            GLOBAL CONFIGURATION                             */
/* -------------------------------------------------------------------------- */

// The lock has positions from 0 to 99 (100 total positions).
// We start at position 50.
let lockCurrentState = 50;

// The size of the circular lock.
// When we move past 99, we wrap back to 0.
const LOCK_WRAP = 100;


/* -------------------------------------------------------------------------- */
/*                          MOVEMENT / TURN FUNCTIONS                          */
/* -------------------------------------------------------------------------- */

/**
 * Turn the lock LEFT by `n` clicks.
 *
 * This function does NOT mutate state.
 * It simply calculates and returns the new position.
 *
 * Why the weird math?
 * - JavaScript's `%` operator can return negative numbers.
 * - We normalize it so the result is ALWAYS between 0 and 99.
 */
function leftTurn(state, n) {
  // Example:
  // state = 10, n = 20
  // 10 - 20 = -10
  // (-10 % 100) = -10  ❌
  // (-10 + 100) = 90   ✅
  // 90 % 100 = 90
  return ((state - n) % LOCK_WRAP + LOCK_WRAP) % LOCK_WRAP;
}

/**
 * Turn the lock RIGHT by `n` clicks.
 *
 * This is simpler because positive modulo works as expected.
 */
function rightTurn(state, n) {
  // Example:
  // state = 90, n = 15
  // 90 + 15 = 105
  // 105 % 100 = 5
  return (state + n) % LOCK_WRAP;
}


/* -------------------------------------------------------------------------- */
/*                          WRAP (BOUNDARY CROSSING) LOGIC                     */
/* -------------------------------------------------------------------------- */

/**
 * Counts how many times we CROSS position 0
 * during a movement.
 *
 * Important:
 * - This does NOT simulate step-by-step movement.
 * - It uses math to compute wraps in O(1) time.
 *
 * @param {number} value - starting position (0–99)
 * @param {number} delta - signed movement (+ for right, - for left)
 * @returns {number} number of times we wrapped around 0
 */
function wraps(value, delta) {
  // Temporary unbounded position (no modulo applied yet)
  const tmp = value + delta;
  let clicks;

  if (delta > 0) {
    /**
     * RIGHT TURN LOGIC
     *
     * We compare how many full 100s fit before and after moving.
     * If the quotient increases, we crossed 0.
     */
    clicks =
      Math.floor(tmp / LOCK_WRAP) -
      Math.floor(value / LOCK_WRAP);
  } else {
    /**
     * LEFT TURN LOGIC
     *
     * Negative division is tricky.
     * Subtracting 1 shifts the boundary so we don't
     * falsely count when landing exactly on 0.
     */
    clicks =
      Math.floor((value - 1) / LOCK_WRAP) -
      Math.floor((tmp - 1) / LOCK_WRAP);
  }

  // Wrap count must always be positive
  return Math.abs(clicks);
}


/* -------------------------------------------------------------------------- */
/*                                  MAIN LOGIC                                 */
/* -------------------------------------------------------------------------- */

function main() {

  // Part 1:
  // How many times we LAND exactly on position 0
  let zeroCounter = 0;

  // Part 2:
  // How many times we CROSS position 0
  let wrapCounter = 0;

  /**
   * We copy the global state into a local variable.
   *
   * Why?
   * - Local variables are faster
   * - Avoids accidental mutation during the loop
   */
  let currentState = lockCurrentState;

  let buffer;
  try {
    /**
     * Read the file as ONE string.
     *
     * We intentionally avoid splitting immediately:
     * - Splitting creates many small strings
     * - This puts pressure on the Garbage Collector
     */
    buffer = readFileSync("../inputs/input-2025.txt", "utf8");
  } catch (err) {
    console.error("\nError: 'input-2025.txt' not found.");
    return;
  }

  /**
   * Split the input into lines.
   * Using a simple array and index-based loop
   * is faster than for...of on large inputs.
   */
  const lines = buffer.split("\n");
  const totalLines = lines.length;

  for (let i = 0; i < totalLines; i++) {
    const rawLine = lines[i];

    // Lines shorter than 2 characters can't be valid commands
    // Example valid line: "R15"
    if (rawLine.length < 2) continue;

    /**
     * We avoid .trim() and .toUpperCase():
     * - Index access is faster
     * - Input format is trusted
     */
    const direction = rawLine[0];
    const n = parseInt(rawLine.slice(1), 10);

    // Skip malformed lines
    if (isNaN(n)) continue;

    /* --------------------------- PART 2: WRAPS --------------------------- */

    // Convert direction to signed movement
    const delta =
      direction === "R" || direction === "r" ? n : -n;

    // Count how many times we crossed position 0
    wrapCounter += wraps(currentState, delta);

    /* -------------------------- PART 1: MOVEMENT -------------------------- */

    // Use direct branching instead of function maps (faster)
    if (delta > 0) {
      currentState = rightTurn(currentState, n);
    } else {
      currentState = leftTurn(currentState, n);
    }

    /* ---------------------------- FINAL CHECK ----------------------------- */

    // Count exact landings on 0
    if (currentState === 0) {
      zeroCounter += 1;
    }
  }

  // Persist the final state back to the global variable
  lockCurrentState = currentState;

  console.log(`Part 1 Answer: ${zeroCounter}`);
  console.log(`Part 2 Answer: ${wrapCounter}`);
}

/* -------------------------------------------------------------------------- */
/*                                   EXECUTION                                 */
/* -------------------------------------------------------------------------- */

console.time("Performance");
main();
console.timeEnd("Performance");
